File: typescript/test/resource.test.ts
import { readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { describe, expect, it } from "vitest";

import { AccessTokenValidator, ProtectedResource, bearerToken, errors } from "../src/validator/index.js";

const VALIDATION = resolve(dirname(fileURLToPath(import.meta.url)), "..", "..", "validation");
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const vectors: any = JSON.parse(readFileSync(join(VALIDATION, "tokens.json"), "utf8"));

function resource(): ProtectedResource {
  const validator = new AccessTokenValidator({
    issuer: vectors.issuer,
    audiences: [vectors.audience],
    jwks: vectors.jwks,
    requiredScopes: vectors.required_scopes,
    acceptedAlgorithms: vectors.accepted_algorithms,
  });
  return new ProtectedResource(vectors.audience, [vectors.issuer], validator, ["read", "write"]);
}

function tokenNamed(name: string): string {
  return vectors.cases.find((c: { name: string }) => c.name === name).token;
}

describe("protected resource", () => {
  it("extracts bearer tokens", () => {
    expect(bearerToken("Bearer abc.def")).toBe("abc.def");
    expect(bearerToken("bearer abc")).toBe("abc");
    expect(bearerToken("Basic abc")).toBeNull();
    expect(bearerToken("")).toBeNull();
    expect(bearerToken(null)).toBeNull();
    expect(bearerToken("Bearer ")).toBeNull();
  });

  it("emits RFC 9728 metadata and paths", () => {
    const pr = resource();
    const md = pr.metadata();
    expect(md.resource).toBe("https://api.example.com");
    expect(md.authorization_servers).toEqual(["https://issuer.example.com"]);
    expect(md.bearer_methods_supported).toEqual(["header"]);
    expect(md.scopes_supported).toEqual(["read", "write"]);
    expect(pr.metadataPath()).toBe("/.well-known/oauth-protected-resource");
    expect(pr.metadataUrl()).toBe("https://api.example.com/.well-known/oauth-protected-resource");

    const withPath = new ProtectedResource(
      "https://mcp.example.com/mcp",
      [vectors.issuer],
      new AccessTokenValidator({ issuer: vectors.issuer, audiences: ["https://mcp.example.com/mcp"], jwks: vectors.jwks }),
    );
    expect(withPath.metadataPath()).toBe("/.well-known/oauth-protected-resource/mcp");
  });

  it("authorizes a valid token", async () => {
    const d = await resource().authenticate("Bearer " + tokenNamed("valid"));
    expect(d.authorized).toBe(true);
    expect(d.status).toBe(200);
    expect(d.result?.subject).toBe("agent-1");
  });

  it("challenges a missing token with 401", async () => {
    const d = await resource().authenticate(null);
    expect(d.authorized).toBe(false);
    expect(d.status).toBe(401);
    expect(d.wwwAuthenticate).toContain(
      'resource_metadata="https://api.example.com/.well-known/oauth-protected-resource"',
    );
  });

  it("maps an expired token to 401 invalid_token", async () => {
    const d = await resource().authenticate("Bearer " + tokenNamed("expired"));
    expect(d.status).toBe(401);
    expect(d.error).toBe("invalid_token");
    expect(d.wwwAuthenticate).toContain('error="invalid_token"');
  });

  it("maps a wrong-audience token to 401", async () => {
    const d = await resource().authenticate("Bearer " + tokenNamed("wrong_audience"));
    expect(d.status).toBe(401);
    expect(d.result?.error).toBe(errors.INVALID_AUDIENCE);
  });

  it("maps insufficient scope to 403", async () => {
    const d = await resource().authenticate("Bearer " + tokenNamed("valid"), ["read", "admin"]);
    expect(d.status).toBe(403);
    expect(d.error).toBe("insufficient_scope");
    expect(d.wwwAuthenticate).toContain('error="insufficient_scope"');
  });
});

File: go/tokenvalidator/resource_test.go
package tokenvalidator

import (
	"encoding/json"
	"strings"
	"testing"
)

func resourceFixture(t *testing.T) *ProtectedResource {
	t.Helper()
	v := loadVectors(t)
	validator, err := New(&Config{
		Issuer: v.Issuer, Audiences: []string{v.Audience}, JWKS: v.JWKS,
		RequiredScopes: v.RequiredScopes, AcceptedAlgorithms: v.AcceptedAlgorithms,
	})
	if err != nil {
		t.Fatalf("new validator: %v", err)
	}
	return NewProtectedResource(v.Audience, []string{v.Issuer}, validator, []string{"read", "write"})
}

func TestBearerToken(t *testing.T) {
	cases := map[string]string{
		"Bearer abc.def": "abc.def",
		"bearer abc":     "abc",
		"Basic abc":      "",
		"":               "",
		"Bearer ":        "",
	}
	for in, want := range cases {
		if got := BearerToken(in); got != want {
			t.Errorf("BearerToken(%q) = %q, want %q", in, got, want)
		}
	}
}

func TestResourceMetadata(t *testing.T) {
	pr := resourceFixture(t)
	md := pr.Metadata()
	if md.Resource != "https://api.example.com" {
		t.Errorf("resource = %q", md.Resource)
	}
	if len(md.AuthorizationServers) != 1 || md.AuthorizationServers[0] != "https://issuer.example.com" {
		t.Errorf("authorization_servers = %v", md.AuthorizationServers)
	}
	if len(md.BearerMethodsSupported) != 1 || md.BearerMethodsSupported[0] != "header" {
		t.Errorf("bearer_methods_supported = %v", md.BearerMethodsSupported)
	}
	if pr.MetadataPath() != "/.well-known/oauth-protected-resource" {
		t.Errorf("MetadataPath = %q", pr.MetadataPath())
	}
	if pr.MetadataURL() != "https://api.example.com/.well-known/oauth-protected-resource" {
		t.Errorf("MetadataURL = %q", pr.MetadataURL())
	}
	b, _ := json.Marshal(md)
	if !strings.Contains(string(b), `"bearer_methods_supported":["header"]`) {
		t.Errorf("metadata JSON = %s", b)
	}
}

func TestMetadataPathWithResourcePath(t *testing.T) {
	v := loadVectors(t)
	validator, _ := New(&Config{Issuer: v.Issuer, Audiences: []string{"https://mcp.example.com/mcp"}, JWKS: v.JWKS})
	pr := NewProtectedResource("https://mcp.example.com/mcp", []string{v.Issuer}, validator, nil)
	if pr.MetadataPath() != "/.well-known/oauth-protected-resource/mcp" {
		t.Errorf("MetadataPath = %q", pr.MetadataPath())
	}
}

func TestAuthenticateValid(t *testing.T) {
	v := loadVectors(t)
	d := resourceFixture(t).Authenticate("Bearer "+tokenNamed(v, "valid"), nil)
	if !d.Authorized || d.Status != 200 {
		t.Fatalf("valid: authorized=%v status=%d", d.Authorized, d.Status)
	}
	if d.Result.Subject != "agent-1" {
		t.Errorf("subject = %q", d.Result.Subject)
	}
}

func TestAuthenticateMissingToken(t *testing.T) {
	d := resourceFixture(t).Authenticate("", nil)
	if d.Authorized || d.Status != 401 {
		t.Fatalf("missing: authorized=%v status=%d", d.Authorized, d.Status)
	}
	if !strings.Contains(d.WWWAuthenticate, `resource_metadata="https://api.example.com/.well-known/oauth-protected-resource"`) {
		t.Errorf("WWW-Authenticate = %q", d.WWWAuthenticate)
	}
}

func TestAuthenticateExpired(t *testing.T) {
	v := loadVectors(t)
	d := resourceFixture(t).Authenticate("Bearer "+tokenNamed(v, "expired"), nil)
	if d.Status != 401 || d.Error != "invalid_token" {
		t.Errorf("expired: status=%d error=%q", d.Status, d.Error)
	}
}

func TestAuthenticateWrongAudience(t *testing.T) {
	v := loadVectors(t)
	d := resourceFixture(t).Authenticate("Bearer "+tokenNamed(v, "wrong_audience"), nil)
	if d.Status != 401 || d.Result.Error != ErrInvalidAudience {
		t.Errorf("wrong audience: status=%d error=%q", d.Status, d.Result.Error)
	}
}

func TestAuthenticateInsufficientScope(t *testing.T) {
	v := loadVectors(t)
	d := resourceFixture(t).Authenticate("Bearer "+tokenNamed(v, "valid"), []string{"read", "admin"})
	if d.Status != 403 || d.Error != "insufficient_scope" {
		t.Errorf("scope: status=%d error=%q", d.Status, d.Error)
	}
	if !strings.Contains(d.WWWAuthenticate, `error="insufficient_scope"`) {
		t.Errorf("WWW-Authenticate = %q", d.WWWAuthenticate)
	}
}
